src/bms/model/control_flow.rs

```diff
@@ -0,0 +1,631 @@
+            match cur {
+                IfChainEntry::ElseIf { units, next, .. } => {
+                    if idx == 0 {
+                        let old = std::mem::replace(units, new_units.into_iter().collect());
+                        return Some(old);
+                    } else {
+                        idx -= 1;
+                        cur = next.as_mut();
+                    }
+                }
+                IfChainEntry::Else { units } => {
+                    if idx == 0 {
+                        let old = std::mem::replace(units, new_units.into_iter().collect());
+                        return Some(old);
+                    } else {
+                        return None;
+                    }
+                }
+                IfChainEntry::EndIf => return None,
+            }
+        }
+    }
+}
+
+/// If-chain used within a random block.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct IfBlock<'a> {
+    condition: BigUint,
+    head_units: Vec<TokenUnit<'a>>, // units for the initial `#IF` branch
+    chain: IfChainEntry<'a>,        // subsequent `ELSEIF`/`ELSE` nodes
+}
+
+impl<'a> IfBlock<'a> {
+    /// Create a new if-chain with units in the first `#IF` entry.
+    pub fn new_if<U>(cond: BigUint, units: U) -> Self
+    where
+        U: IntoIterator<Item = TokenUnit<'a>>,
+    {
+        Self {
+            condition: cond,
+            head_units: units.into_iter().collect(),
+            chain: IfChainEntry::EndIf,
+        }
+    }
+
+    /// Add an `#ELSEIF <cond>` entry with units.
+    pub fn or_else_if<U>(mut self, cond: BigUint, units: U) -> Self
+    where
+        U: IntoIterator<Item = TokenUnit<'a>>,
+    {
+        let entry = IfChainEntry::ElseIf {
+            cond,
+            units: units.into_iter().collect(),
+            next: Box::new(IfChainEntry::EndIf),
+        };
+        self.chain.append_to_tail(entry);
+        self
+    }
+
+    /// Add an `#ELSE` entry with units.
+    pub fn or_else<U>(mut self, units: U) -> Self
+    where
+        U: IntoIterator<Item = TokenUnit<'a>>,
+    {
+        let entry = IfChainEntry::Else {
+            units: units.into_iter().collect(),
+        };
+        self.chain.append_to_tail(entry);
+        self
+    }
+
+    /// Returns a view of the head `#IF` units.
+    #[must_use]
+    pub fn head_units(&self) -> &[TokenUnit<'a>] {
+        &self.head_units
+    }
+
+    /// Replace units for the branch at `index`. `index=0` refers to the head `#IF` branch;
+    /// `index>=1` refers to the subsequent `ELSEIF`/`ELSE` nodes.
+    /// Returns previous units when the index exists.
+    pub fn set_units_at<U>(&mut self, index: usize, new_units: U) -> Option<Vec<TokenUnit<'a>>>
+    where
+        U: IntoIterator<Item = TokenUnit<'a>>,
+    {
+        if index == 0 {
+            let mut incoming: Vec<TokenUnit<'a>> = new_units.into_iter().collect();
+            std::mem::swap(&mut incoming, &mut self.head_units);
+            Some(incoming)
```

Comment:
It can shorten with `replace`:
```suggestion
            let old = std::mem::replace(&mut self.head_units, new_units.into_iter().collect());
            Some(old)
```